samples/src/main/java/okio/samples/Throttler.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio.samples;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.util.concurrent.TimeUnit;
+import okio.Buffer;
+import okio.ForwardingSink;
+import okio.ForwardingSource;
+import okio.Sink;
+import okio.Source;
+
+/**
+ * Enables throttling of Sources and Sinks. Attach to this throttler via {@link #source(Source)} and
+ * {@link #sink(Sink)} and set the desired throughput via {@link #bytesPerTimePeriod(long, long,
+ * TimeUnit)}. Multiple Sources and Sinks can be attached to a single Throttler and they will be
+ * throttled as a group, where their combined throughput will not exceed the desired throughput. The
+ * same Source or Sink can be attached to multiple Throttlers and its throughput will not exceed the
+ * desired throughput of any of the Throttlers.
+ *
+ * <p>To clear any throttling, use {@link #clear()} or specify a <code>byteCount</code> of <code>
+ * 0</code> when setting the throughput.
+ */
+public final class Throttler {
+  private static final int NANOS_PER_MILLI = 1000000;
+
+  private double limit = 0.0d;
+  private long last = -1;
+
+  /** Sets the desired max throughput. <code>byteCount</code> of zero to clear. */
+  public void bytesPerTimePeriod(long byteCount, long duration, TimeUnit unit) {
+    if (unit == null || byteCount < 0 || duration <= 0) {
+      throw new IllegalArgumentException(
+          "byteCount=" + byteCount + " duration=" + duration + " unit=" + unit);
+    }
+
+    synchronized (this) {
+      this.limit = (double) byteCount / unit.toNanos(duration);
```

Comment:
Nit: I don’t think this is what we want. I think want to actually implement the limit provided and not spread the limit out over the narrowest possible time unit.

In particular, if the user asks for 60 bytes per minute, I want us to:
Send 60 bytes, then sleep 60 seconds, vs.
Send 1 byte, sleep 1 second. 60 times.

Reply:
I use the limit to determine how long a request of a specific `byteCount` should take.

In your example, if the user asks for 180 bytes, the request should take 180 seconds. If the last request was over 180 seconds ago, there will be no waiting, but if it is less, we'll wait for the difference. Then reset the time and process the next request.